src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,258 @@
+                .delete => |delete_ix| {
+                    const cf_index, const key = delete_ix;
+                    self.maps[cf_index].delete(batch.allocator, key);
+                },
+            }
+        }
+    }
+};
+
+pub const MapBatch = struct {
+    allocator: Allocator,
+    instructions: std.ArrayListUnmanaged(Instruction),
+
+    const Instruction = union(enum) {
+        put: struct { usize, []const u8, []const u8 },
+        delete: struct { usize, []const u8 },
+    };
+
+    const Self = @This();
+
+    pub fn init(allocator: Allocator) Self {
+        return .{
+            .allocator = allocator,
+            .instructions = std.ArrayListUnmanaged(Instruction){},
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.instructions.deinit(self.allocator);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+        value: cf.Value,
+    ) !void {
+        const k_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        errdefer self.allocator.free(k_bytes);
+        const v_bytes = try cf.value().serializeAlloc(self.allocator, value);
+        errdefer self.allocator.free(v_bytes);
+        return try self.instructions.append(
+            self.allocator,
+            .{ .put = .{ cf_index, k_bytes, v_bytes } },
+        );
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        const k_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        errdefer self.allocator.free(k_bytes);
+        return try self.instructions.append(
+            self.allocator,
+            .{ .delete = .{ cf_index, k_bytes } },
+        );
+    }
+};
+
+const SharedHashMap = struct {
+    allocator: Allocator,
+    map: std.StringHashMapUnmanaged([]const u8) = .{},
+    lock: DefaultRwLock = .{},
+
+    const Self = @This();
+
+    fn init(allocator: Allocator) Allocator.Error!Self {
+        return .{ .allocator = allocator };
+    }
+
+    pub fn deinit(self: *Self) void {
+        var iter = self.map.iterator();
+        while (iter.next()) |entry| {
+            self.allocator.free(entry.key_ptr.*);
+            self.allocator.free(entry.value_ptr.*);
+        }
+        self.map.deinit(self.allocator);
+    }
+
+    pub fn free(self: Self, bytes: []const u8) void {
+        self.allocator.free(bytes);
+    }
+
+    pub fn put(self: *Self, key: []const u8, value: []const u8) Allocator.Error!void {
+        self.lock.lock();
+        defer self.lock.unlock();
+        try self.map.put(self.allocator, key, value);
+    }
+
+    /// Only call this while holding the lock
+    pub fn getPreLocked(self: *Self, key: []const u8) ?[]const u8 {
+        return self.map.get(key) orelse return null;
+    }
+
+    pub fn delete(self: *Self, liberator: Allocator, key_: []const u8) void {
+        const key, const value = lock: {
```

Comment:
I used a different label here because I wanted to be very clear that the sole purpose of this block is to define the scope of the lock. No code should be put in here unless it needs to run while the lock is held. Is that a good enough reason for a non-standard name? Can you think of a better name for that purpose?